File: Java_client/bim-portal-enhanced/src/main/java/com/pb40/bimportal/examples/OrganizationExample.java
package com.pb40.bimportal.examples;

import com.bimportal.client.api.InfrastrukturApi;
import com.bimportal.client.model.OrganisationForPublicDTO;
import com.pb40.bimportal.auth.AuthService;
import com.pb40.bimportal.auth.AuthServiceImpl;
import com.pb40.bimportal.client.BimPortalClientBuilder;
import com.pb40.bimportal.client.EnhancedBimPortalClient;
import com.pb40.bimportal.config.BimPortalConfig;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Enhanced organization examples for BIM Portal Java client with real JWT user UUID extraction.
 *
 * <p>Demonstrates organization functionality according to the actual Swagger API: - GET
 * /infrastruktur/api/v1/public/organisation (public, no auth required) - GET
 * /infrastruktur/api/v1/public/organisation/my?userId={userId} (requires auth + userId)
 *
 * <p>Now includes real implementation of JWT user UUID extraction from token payload.
 */
public class OrganizationExample {

  private static final Logger logger = LoggerFactory.getLogger(OrganizationExample.class);

  /**
   * Check if credentials are available for authentication.
   *
   * @return True if credentials are configured
   */
  private static boolean checkCredentials() {
    if (!BimPortalConfig.hasCredentials()) {
      System.out.println("============================================================");
      System.out.println("WARNING: Credentials not found in environment variables.");
      System.out.println(
          "Please set "
              + BimPortalConfig.USERNAME_ENV_VAR
              + " and "
              + BimPortalConfig.PASSWORD_ENV_VAR
              + " in .env file.");
      System.out.println("Some organization examples require authentication.");
      System.out.println("============================================================");
      return false;
    }
    return true;
  }

  /**
   * Demonstrate organization API usage with real JWT user UUID extraction.
   *
   * @param client Enhanced BIM Portal client
   */
  public static void runOrganizationExamples(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("🏢 ORGANIZATION API EXAMPLES");
    System.out.println("============================================================");

    System.out.println("\n1️⃣ Fetching all available organizations (public endpoint)...");
    try {
      List<OrganisationForPublicDTO> allOrganizations = client.getAllOrganisations();
      System.out.println("✅ Found " + allOrganizations.size() + " available organizations:");

      for (int i = 0; i < Math.min(5, allOrganizations.size()); i++) {
        OrganisationForPublicDTO org = allOrganizations.get(i);
        System.out.println(
            "   "
                + (i + 1)
                + ". "
                + org.getName()
                + " ("
                + org.getGuid()
                + ")"
                + (org.getDescription() != null ? " - " + org.getDescription() : ""));
      }

      if (allOrganizations.size() > 5) {
        System.out.println("   ... and " + (allOrganizations.size() - 5) + " more");
      }
    } catch (Exception e) {
      System.err.println("❌ Error fetching organizations: " + e.getMessage());
      logger.error("Error fetching organizations", e);
    }

    System.out.println("\n2️⃣ Testing organization search functionality...");
    try {
      List<OrganisationForPublicDTO> allOrgs = client.getAllOrganisations();
      if (!allOrgs.isEmpty()) {
        OrganisationForPublicDTO firstOrg = allOrgs.get(0);

        // Test GUID search
        Optional<OrganisationForPublicDTO> foundByGuid =
            client.findOrganisationByGuid(firstOrg.getGuid().toString());
        System.out.println(
            "   🔍 Search by GUID: " + (foundByGuid.isPresent() ? "✅ Found" : "❌ Not found"));

        // Test name search
        if (firstOrg.getName() != null) {
          Optional<OrganisationForPublicDTO> foundByName =
              client.findOrganisationByName(firstOrg.getName());
          System.out.println(
              "   🔍 Search by name: " + (foundByName.isPresent() ? "✅ Found" : "❌ Not found"));
        }

        // Test availability check
        boolean hasOrgs = client.hasAvailableOrganisations();
        System.out.println("   📊 Organizations available: " + (hasOrgs ? "✅ Yes" : "❌ No"));
      }
    } catch (Exception e) {
      System.err.println("❌ Error in organization search tests: " + e.getMessage());
      logger.error("Error in organization search tests", e);
    }

    System.out.println("\n3️⃣ Testing JWT user UUID extraction and user organizations...");
    if (!client.isAuthenticated()) {
      System.out.println("   ⚠️  Authentication required for user organization endpoints");
      System.out.println("   💡 Client is not authenticated - skipping user organization examples");
    } else {
      System.out.println("   ✅ Client is authenticated");

      // Extract user UUID from JWT token
      demonstrateRealUserOrganizationMethods(client);
    }

    System.out.println("\n4️⃣ Testing query parameters approach with real user UUID...");
    if (client.isAuthenticated()) {
      try {
        // Get the auth service to access user UUID
        AuthService authService = client.getAuthService();
        if (authService instanceof AuthServiceImpl) {
          AuthServiceImpl authServiceImpl = (AuthServiceImpl) authService;

          Optional<UUID> currentUserId = authServiceImpl.getCurrentUserId();
          if (currentUserId.isPresent()) {
            UUID userId = currentUserId.get();
            InfrastrukturApi.GetOrganisationsOfUserQueryParams queryParams =
                client.createUserOrganisationQueryParams(userId);

            System.out.println("   ✅ Query parameters created for real user: " + userId);
            System.out.println(
                "   💡 Ready to call: client.getUserOrganisationsWithParams(queryParams)");

            // Actually test the query parameters
            try {
              List<OrganisationForPublicDTO> userOrgs =
                  client.getUserOrganisationsWithParams(queryParams);
              System.out.println(
                  "   ✅ User organizations retrieved: " + userOrgs.size() + " found");
            } catch (Exception e) {
              System.out.println("   ⚠️  User organizations call failed: " + e.getMessage());
              logger.debug("User organizations API call failed", e);
            }
          } else {
            System.out.println("   ⚠️  User UUID not available from JWT token");
            System.out.println(
                "   💡 Check JWT token structure - user ID may be in different claim");
          }
        }
      } catch (Exception e) {
        System.err.println("   ❌ Error in query parameters test: " + e.getMessage());
        logger.error("Error in query parameters test", e);
      }
    } else {
      System.out.println("   ⚠️  Authentication required for real user UUID extraction");
    }
  }

  /** Demonstrate user organization methods with real JWT user UUID extraction. */
  private static void demonstrateRealUserOrganizationMethods(EnhancedBimPortalClient client) {
    System.out.println("\n   🔐 Real JWT User UUID Extraction:");

    try {
      // Get the auth service to access user UUID
      AuthService authService = client.getAuthService();
      if (!(authService instanceof AuthServiceImpl)) {
        System.out.println("   ⚠️  AuthService implementation not available for UUID extraction");
        return;
      }

      AuthServiceImpl authServiceImpl = (AuthServiceImpl) authService;

      // Extract user UUID from JWT token
      Optional<UUID> currentUserId = authServiceImpl.getCurrentUserId();

      if (currentUserId.isEmpty()) {
        System.out.println("   ❌ User UUID not available from JWT token");
        System.out.println("   💡 Possible reasons:");
        System.out.println("      • JWT token doesn't contain user ID claim");
        System.out.println("      • User ID claim is in unexpected format");
        System.out.println("      • Token parsing failed");

        // Show token status for debugging
        String tokenStatus = authServiceImpl.getTokenStatus();
        System.out.println("   🔍 Token status: " + tokenStatus);
        return;
      }

      UUID userId = currentUserId.get();
      System.out.println("   ✅ Successfully extracted user UUID: " + userId);

      // Now test all user organization methods with real UUID
      System.out.println("\n   📋 Testing User Organization Methods:");

      try {
        // Test 1: Get user organizations
        System.out.println("   1️⃣ Getting user organizations...");
        List<OrganisationForPublicDTO> userOrgs = client.getUserOrganisations(userId);
        System.out.println("   ✅ Found " + userOrgs.size() + " organizations for user");

        for (int i = 0; i < Math.min(3, userOrgs.size()); i++) {
          OrganisationForPublicDTO org = userOrgs.get(i);
          System.out.println("      👤 " + org.getName() + " (" + org.getGuid() + ")");
        }

        if (userOrgs.size() > 3) {
          System.out.println("      ... and " + (userOrgs.size() - 3) + " more");
        }

        // Test 2: Organization count
        System.out.println("\n   2️⃣ Getting organization count...");
        int orgCount = client.getUserOrganisationsCount(userId);
        System.out.println("   ✅ User is member of " + orgCount + " organizations");

        // Test 3: Organization names
        System.out.println("\n   3️⃣ Getting organization names...");
        List<String> orgNames = client.getUserOrganisationNames(userId);
        System.out.println("   ✅ Organization names: " + String.join(", ", orgNames));

        // Test 4: Membership check (if user has organizations)
        if (!userOrgs.isEmpty()) {
          System.out.println("\n   4️⃣ Testing membership check...");
          OrganisationForPublicDTO firstOrg = userOrgs.get(0);
          boolean isMember = client.isUserMemberOfOrganisation(userId, firstOrg.getGuid());
          System.out.println(
              "   ✅ Membership check for '"
                  + firstOrg.getName()
                  + "': "
                  + (isMember ? "✅ Confirmed" : "❌ Not confirmed"));
        } else {
          System.out.println("\n   4️⃣ Skipping membership check (no organizations found)");
        }

        // Test 5: First organization (convenience method)
        System.out.println("\n   5️⃣ Getting first user organization...");
        Optional<OrganisationForPublicDTO> firstUserOrg = client.getUserFirstOrganisation(userId);
        if (firstUserOrg.isPresent()) {
          System.out.println("   ✅ First organization: " + firstUserOrg.get().getName());
        } else {
          System.out.println("   ⚠️  No organizations found for user");
        }

      } catch (Exception e) {
        System.err.println("   ❌ Error in user organization API calls: " + e.getMessage());
        logger.error("Error in user organization API calls", e);

        // Provide debugging info
        if (e.getMessage().contains("404") || e.getMessage().contains("Not Found")) {
          System.out.println(
              "   💡 404 error suggests the user organization endpoint may not be available");
          System.out.println("      or the user UUID format is not accepted by the API");
        } else if (e.getMessage().contains("401") || e.getMessage().contains("Unauthorized")) {
          System.out.println("   💡 401 error suggests authentication issues");
        } else if (e.getMessage().contains("403") || e.getMessage().contains("Forbidden")) {
          System.out.println(
              "   💡 403 error suggests insufficient permissions for user organization access");
        }
      }

    } catch (Exception e) {
      System.err.println("   ❌ Error in JWT user UUID extraction: " + e.getMessage());
      logger.error("Error in JWT user UUID extraction", e);
    }
  }

  /** Demonstrate the JWT token structure analysis for debugging. */
  public static void demonstrateJwtTokenAnalysis(EnhancedBimPortalClient client) {
    System.out.println("\n============================================================");
    System.out.println("🔍 JWT TOKEN ANALYSIS FOR DEBUGGING");
    System.out.println("============================================================");

    if (!client.isAuthenticated()) {
      System.out.println("❌ Not authenticated - cannot analyze JWT token");
      return;
    }

    try {
      AuthService authService = client.getAuthService();
      if (authService instanceof AuthServiceImpl) {
        AuthServiceImpl authServiceImpl = (AuthServiceImpl) authService;

        String tokenStatus = authServiceImpl.getTokenStatus();
        System.out.println("📊 Token Status: " + tokenStatus);

        // Get the current token for analysis
        String currentToken = authServiceImpl.getValidToken();
        if (currentToken != null) {
          System.out.println("✅ Current token available for analysis");

          // Analyze token structure
          analyzeJwtTokenStructure(currentToken);

          // Test user ID extraction
          Optional<UUID> extractedUserId = authServiceImpl.getCurrentUserId();
          if (extractedUserId.isPresent()) {
            System.out.println("✅ User UUID extraction: " + extractedUserId.get());
          } else {
            System.out.println("❌ User UUID extraction failed");
            System.out.println("💡 Check the JWT payload claims for user identification");
          }
        } else {
          System.out.println("❌ No valid token available for analysis");
        }
      }
    } catch (Exception e) {
      System.err.println("❌ Error in JWT token analysis: " + e.getMessage());
      logger.error("Error in JWT token analysis", e);
    }
  }

  /**
   * Analyze JWT token structure for debugging purposes.
   *
   * @param token JWT token string
   */
  private static void analyzeJwtTokenStructure(String token) {
    try {
      System.out.println("\n🔬 JWT Token Structure Analysis:");

      String[] parts = token.split("\\.");
      System.out.println("   📋 Token parts: " + parts.length + " (expected: 3 for JWT)");

      if (parts.length >= 2) {
        // Decode and display payload
        String payload = new String(java.util.Base64.getUrlDecoder().decode(parts[1]));
        System.out.println(
            "   📄 JWT Payload (first 200 chars): "
                + payload.substring(0, Math.min(200, payload.length()))
                + "...");

        // Look for common user ID claims
        System.out.println("   🔍 Checking for user ID claims:");
        checkForClaim(payload, "sub");
        checkForClaim(payload, "userId");
        checkForClaim(payload, "user_id");
        checkForClaim(payload, "id");
        checkForClaim(payload, "email");
        checkForClaim(payload, "username");
      }

    } catch (Exception e) {
      System.err.println("   ❌ Failed to analyze token structure: " + e.getMessage());
    }
  }

  /**
   * Check if a specific claim exists in the JWT payload.
   *
   * @param payload JWT payload string
   * @param claimName Name of the claim to check
   */
  private static void checkForClaim(String payload, String claimName) {
    if (payload.contains("\"" + claimName + "\"")) {
      try {
        // Simple regex to extract claim value
        java.util.regex.Pattern pattern =
            java.util.regex.Pattern.compile("\"" + claimName + "\"\\s*:\\s*\"?([^,}\"]+)\"?");
        java.util.regex.Matcher matcher = pattern.matcher(payload);
        if (matcher.find()) {
          String value = matcher.group(1);
          System.out.println("      ✅ " + claimName + ": " + value);
        } else {
          System.out.println("      ⚠️  " + claimName + ": found but value extraction failed");
        }
      } catch (Exception e) {
        System.out.println("      ❌ " + claimName + ": error extracting value");
      }
    } else {
      System.out.println("      ❌ " + claimName + ": not found");
    }
  }

  /** Main method to run enhanced organization examples with real JWT user UUID extraction. */
  public static void main(String[] args) {
    System.out.println("======================================================================");
    System.out.println("🚀 ENHANCED BIM PORTAL ORGANIZATION API EXAMPLES");
    System.out.println("   📋 Features: Real JWT User UUID Extraction");
    System.out.println("======================================================================");

    boolean hasCredentials = checkCredentials();

    System.out.println("🔧 Setting up client...");

    try {
      EnhancedBimPortalClient client = BimPortalClientBuilder.buildDefault();

      // Health check
      EnhancedBimPortalClient.HealthCheckResult healthCheck = client.performHealthCheck();
      System.out.println("🩺 Health check: " + healthCheck);

      if (!healthCheck.isApiAccessible()) {
        System.err.println("❌ API is not accessible. Please check configuration.");
        return;
      }

      // Run organization examples
      runOrganizationExamples(client);

      // Run JWT token analysis for debugging
      if (hasCredentials && client.isAuthenticated()) {
        demonstrateJwtTokenAnalysis(client);
      }

      System.out.println(
          "\n======================================================================");
      System.out.println("✅ ENHANCED ORGANIZATION EXAMPLES COMPLETE!");
      System.out.println("======================================================================");

      if (hasCredentials && client.isAuthenticated()) {
        System.out.println("🎯 Key achievements:");
        System.out.println("   ✅ JWT token user UUID extraction implemented");
        System.out.println("   ✅ Real user organization API calls demonstrated");
        System.out.println("   ✅ Comprehensive error handling and debugging");

        AuthService authService = client.getAuthService();
        if (authService instanceof AuthServiceImpl) {
          AuthServiceImpl authServiceImpl = (AuthServiceImpl) authService;
          Optional<UUID> userId = authServiceImpl.getCurrentUserId();
          if (userId.isPresent()) {
            System.out.println("   ✅ Your user UUID: " + userId.get());
          }
        }
      } else {
        System.out.println("💡 Set up credentials to test authenticated organization endpoints");
        System.out.println("   and experience JWT user UUID extraction");
      }

      // Cleanup
      if (client.isAuthenticated()) {
        client.logout();
        System.out.println("👋 Logged out and session closed.");
      }

    } catch (Exception e) {
      logger.error("❌ Error running enhanced organization examples", e);
      System.err.println("❌ Error running examples: " + e.getMessage());
    }
  }
}

File: Python_client/examples/organization_example.py
"""
Organization examples for BIM Portal Python client.

This module demonstrates organization functionality including:
- Getting all organizations (public endpoint)
- Getting user organizations (authenticated)
- Organization search and filtering capabilities
"""

import os
import sys
from pathlib import Path

# Ensure we can import from project root
project_root = Path(__file__).parent.parent
sys.path.insert(0, str(project_root))

import logging
from typing import List, Optional
from uuid import UUID

from dotenv import load_dotenv

from client.auth.auth_config import BIM_PORTAL_PASSWORD_ENV_VAR, BIM_PORTAL_USERNAME_ENV_VAR
from client.auth.auth_service_impl import AuthService
from client.enhanced_bim_client import EnhancedBimPortalClient
from client.config import BIMPortalConfig

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Load environment variables
load_dotenv()


def check_credentials() -> bool:
    """
    Check if credentials are available for authentication.
    
    Returns:
        True if credentials are configured
    """
    if not os.getenv(BIM_PORTAL_USERNAME_ENV_VAR) or not os.getenv(BIM_PORTAL_PASSWORD_ENV_VAR):
        print("=" * 60)
        print("WARNING: Credentials not found in environment variables.")
        print(f"Please set {BIM_PORTAL_USERNAME_ENV_VAR} and {BIM_PORTAL_PASSWORD_ENV_VAR} in .env file.")
        print("Some organization examples require authentication.")
        print("=" * 60)
        return False
    return True


def run_organization_examples(client: EnhancedBimPortalClient):
    """
    Demonstrate organization API usage.
    
    Args:
        client: Enhanced BIM Portal client
    """
    print("\n" + "=" * 60)
    print("🏢 ORGANIZATION API EXAMPLES")
    print("=" * 60)
    
    print("\n1️⃣ Fetching all available organizations (public endpoint)...")
    try:
        all_organizations = client.get_organisations()
        print(f"✅ Found {len(all_organizations)} available organizations:")
        
        for i, org in enumerate(all_organizations[:5], 1):
            description = f" - {org.description}" if hasattr(org, 'description') and org.description else ""
            print(f"   {i}. {org.name} ({org.guid}){description}")
        
        if len(all_organizations) > 5:
            print(f"   ... and {len(all_organizations) - 5} more")
            
    except Exception as e:
        print(f"❌ Error fetching organizations: {e}")
        logger.error("Error fetching organizations", exc_info=True)
    
    print("\n2️⃣ Testing organization search functionality...")
    try:
        all_orgs = client.get_organisations()
        if all_orgs:
            first_org = all_orgs[0]
            
            # Test GUID search
            found_by_guid = find_organization_by_guid(all_orgs, first_org.guid)
            print(f"   🔍 Search by GUID: {'✅ Found' if found_by_guid else '❌ Not found'}")
            
            # Test name search
            if hasattr(first_org, 'name') and first_org.name:
                found_by_name = find_organization_by_name(all_orgs, first_org.name)
                print(f"   🔍 Search by name: {'✅ Found' if found_by_name else '❌ Not found'}")
            
            # Test availability check
            has_orgs = len(all_orgs) > 0
            print(f"   📊 Organizations available: {'✅ Yes' if has_orgs else '❌ No'}")
            
    except Exception as e:
        print(f"❌ Error in organization search tests: {e}")
        logger.error("Error in organization search tests", exc_info=True)
    
    print("\n3️⃣ Testing user organizations (requires authentication)...")
    if not is_authenticated(client):
        print("   ⚠️  Authentication required for user organization endpoints")
        print("   💡 Client is not authenticated - skipping user organization examples")
    else:
        print("   ✅ Client is authenticated")
        demonstrate_user_organization_methods(client)


def find_organization_by_guid(organizations: List, guid: UUID) -> Optional[any]:
    """Find organization by GUID."""
    for org in organizations:
        if org.guid == guid:
            return org
    return None


def find_organization_by_name(organizations: List, name: str) -> Optional[any]:
    """Find organization by name."""
    for org in organizations:
        if hasattr(org, 'name') and org.name == name:
            return org
    return None


def is_authenticated(client: EnhancedBimPortalClient) -> bool:
    """Check if client is authenticated."""
    try:
        # Try to get a valid token to check authentication
        token = client.auth_service.get_valid_token()
        return token is not None
    except Exception:
        return False


def demonstrate_user_organization_methods(client: EnhancedBimPortalClient):
    """Demonstrate user organization methods."""
    print("\n   🔍 User Organization Methods:")
    
    try:
        # Get user organizations
        print("   1️⃣ Getting user organizations...")
        user_orgs = client.get_my_organisations()
        print(f"   ✅ Found {len(user_orgs)} organizations for user")
        
        for i, org in enumerate(user_orgs[:3], 1):
            print(f"      👤 {org.name} ({org.guid})")
        
        if len(user_orgs) > 3:
            print(f"      ... and {len(user_orgs) - 3} more")
        
        # Organization count
        print(f"\n   2️⃣ Organization count: {len(user_orgs)} organizations")
        
        # Organization names
        print("\n   3️⃣ Getting organization names...")
        org_names = [org.name for org in user_orgs if hasattr(org, 'name')]
        print(f"   ✅ Organization names: {', '.join(org_names)}")
        
        # First organization (convenience method)
        print("\n   4️⃣ Getting first user organization...")
        if user_orgs:
            first_user_org = user_orgs[0]
            print(f"   ✅ First organization: {first_user_org.name}")
        else:
            print("   ⚠️  No organizations found for user")
            
    except Exception as e:
        print(f"   ❌ Error in user organization API calls: {e}")
        logger.error("Error in user organization API calls", exc_info=True)
        
        # Provide debugging info
        error_msg = str(e).lower()
        if "404" in error_msg or "not found" in error_msg:
            print("   💡 404 error suggests the user organization endpoint may not be available")
        elif "401" in error_msg or "unauthorized" in error_msg:
            print("   💡 401 error suggests authentication issues")
        elif "403" in error_msg or "forbidden" in error_msg:
            print("   💡 403 error suggests insufficient permissions for user organization access")


def demonstrate_jwt_token_analysis(client: EnhancedBimPortalClient):
    """Demonstrate JWT token analysis for debugging."""
    print("\n" + "=" * 60)
    print("🔍 JWT TOKEN ANALYSIS FOR DEBUGGING")
    print("=" * 60)
    
    if not is_authenticated(client):
        print("❌ Not authenticated - cannot analyze JWT token")
        return
    
    try:
        # Get current token for analysis
        current_token = client.auth_service.get_valid_token()
        if current_token:
            print("✅ Current token available for analysis")
            
            # Analyze token structure
            analyze_jwt_token_structure(current_token)
            
        else:
            print("❌ No valid token available for analysis")
            
    except Exception as e:
        print(f"❌ Error in JWT token analysis: {e}")
        logger.error("Error in JWT token analysis", exc_info=True)


def analyze_jwt_token_structure(token: str):
    """Analyze JWT token structure for debugging purposes."""
    try:
        import base64
        import json
        
        print("\n🔬 JWT Token Structure Analysis:")
        
        parts = token.split(".")
        print(f"   📋 Token parts: {len(parts)} (expected: 3 for JWT)")
        
        if len(parts) >= 2:
            # Decode and display payload
            # Add padding if needed for base64 decoding
            payload_b64 = parts[1]
            payload_b64 += "=" * (4 - len(payload_b64) % 4)
            
            try:
                payload_bytes = base64.urlsafe_b64decode(payload_b64)
                payload = payload_bytes.decode('utf-8')
                payload_json = json.loads(payload)
                
                print(f"   📄 JWT Payload preview: {str(payload_json)[:200]}...")
                
                # Look for common user ID claims
                print("   🔍 Checking for user ID claims:")
                check_for_claim(payload_json, "sub") # This is the user id
                check_for_claim(payload_json, "email") # This is the username

                
            except Exception as decode_error:
                print(f"   ❌ Failed to decode payload: {decode_error}")
                
    except Exception as e:
        print(f"   ❌ Failed to analyze token structure: {e}")


def check_for_claim(payload_json: dict, claim_name: str):
    """Check if a specific claim exists in the JWT payload."""
    if claim_name in payload_json:
        value = payload_json[claim_name]
        print(f"      ✅ {claim_name}: {value}")
    else:
        print(f"      ❌ {claim_name}: not found")


def main():
    """Main method to run organization examples."""
    print("=" * 70)
    print("🚀 BIM PORTAL ORGANIZATION API EXAMPLES")
    print("   📋 Features: Organization Management & JWT Analysis")
    print("=" * 70)
    
    has_credentials = check_credentials()
    
    print("🔧 Setting up client...")
    
    try:
        auth_service = AuthService()
        client = EnhancedBimPortalClient(
            auth_service=auth_service,
            base_url=BIMPortalConfig.BASE_URL
        )
        
        # Run organization examples
        run_organization_examples(client)
        
        # Run JWT token analysis for debugging (if authenticated)
        if has_credentials and is_authenticated(client):
            demonstrate_jwt_token_analysis(client)
        
        print("\n" + "=" * 70)
        print("✅ ORGANIZATION EXAMPLES COMPLETE!")
        print("=" * 70)
        
        if has_credentials and is_authenticated(client):
            print("🎯 Key achievements:")
            print("   ✅ Organization API endpoints demonstrated")
            print("   ✅ User organization access tested")
            print("   ✅ JWT token analysis performed")
        else:
            print("💡 Set up credentials to test authenticated organization endpoints")
            print("   and experience JWT token analysis")
        
    except Exception as e:
        logger.error("Error running organization examples", exc_info=True)
        print(f"❌ Error running examples: {e}")


if __name__ == "__main__":
    main()
